url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/syntax/builder/ArtificialRule.java

```diff
@@ -0,0 +1,228 @@
+package fri.patterns.interpreter.parsergenerator.syntax.builder;
+
+import java.util.*;
+import fri.patterns.interpreter.parsergenerator.Token;
+import fri.patterns.interpreter.parsergenerator.util.SymbolToName;
+
+/**
+	ArtificialRule is needed to
+	<ul>
+		<li>create nonterminals and rules for symbols within parenthesis: "(a b c)" -> "_a_b_c_"</li>
+		<li>create nonterminals and rules for symbols that were quantified with "*", "+", "?"</li>
+	</ul>
+	The nonterminal names get created from symbols that are converted to names.
+	Every created nonterminal starts with "_" (Token.ARTIFICIAL_RULE_START_CHARACTER).
+	<p>
+	Example:
+	<pre>
+		sentence1 ::= word* ;
+		sentence2 ::= word+ ;
+		sentence3 ::= word? ;
+	</pre>
+	will be converted to following rules
+	<pre>
+		_sentence1_OPTLIST ::= _sentence1_LIST;
+		_sentence1_OPTLIST ::= ;
+		_sentence1_LIST ::= _sentence1_LIST word;
+		_sentence1_LIST ::= word;
+		
+		_sentence2_LIST ::= _sentence2_LIST word;
+		_sentence2_LIST ::= word;
+
+		_sentence3_OPT ::= word;
+		_sentence3_OPT ::= ;
+	</pre>
+	 
+	@author (c) 2002 Fritz Ritzberger
+*/
+
+class ArtificialRule
+{
+	private List rules;
+	private String nonterminal;
+	
+	/**
+		Creates an artificial rule from an expression within parenthesis.
+		Argument sentencesInParenthesis holds a List of arbitrary deep Lists
+		that contain String or ArtificialRule at end.
+		The method getRules() will return as much rules as are contained on any
+		level in sentencesInParenthesis.
+	*/
+	public ArtificialRule(List sentencesInParenthesis, String catSym)	{	// make "(b c)" to rule "_b_c_ ::= b c"
+		StringBuffer sb = new StringBuffer();
+		parenthesisContentsToString(sentencesInParenthesis, sb, catSym);
+		nonterminal = ensureUnderscore(sb.toString());
+		rules = createRule(nonterminal, sentencesInParenthesis);
+	}
+	
+	private void parenthesisContentsToString(List sentences, StringBuffer sb, String catSym)	{
+		for (int i = 0; i < sentences.size(); i++)	{
+			Object o = sentences.get(i);
+			
+			if (o instanceof List)	{
+				parenthesisContentsToString((List)o, sb, "");	// catenize symbol only on first level
+			}
+			else	{
+				if (o instanceof String)
+					o = SymbolToName.makeIdentifier((String)o, "");	// no "_" for quotes
+				sb.append(o.toString());	// ArtificialRule goes here
+			}
+
+			if (i < sentences.size() - 1)
+				sb.append((catSym.length() > 0 ? "_" + catSym : "") + "_");
+		}
+	}
+
+	
+	/**
+		Creates an artificial rule from with a quantifier like "*", "+", "?".
+		From this some rules are resulting that getRules() will return.
+		"token" is either String or ArtificialRule.
+	*/
+	public ArtificialRule(Object token, String quantifier)	{	// make "a*", "a+", "a?" to rules
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 108-130 into a separate method.